### ISSUE TITLE
HTTP response body not sent (javaagent mode)

### ISSUE THREAD

**Issue by akira-kurogane** [prometheus/jmx_exporter#718]
Hi, I'm using jmx_exporter v0.17.0 in javaagent mode.

With one client I want to use, a splunk data input module, the body of the HTTP response fails to be sent.

HTTP request:
```
GET /metrics HTTP/1.1
Host: kfk-akira-1.asia-northeast1-b.c.test-stretched-cluster.internal:8297
User-Agent: Go-http-client/1.1
Accept-Encoding: gzip
```
The HTTP response:
```
HTTP/1.1 200 OK
Content-encoding: gzip
Date: Wed, 29 Jun 2022 02:11:14 GMT
Transfer-encoding: chunked
Content-type: text/plain; version=0.0.4; charset=utf-8

<JUST STOPS HERE - no body/chunk size and no data>
```
It is only this one client that I can reproduce the issue with (https://github.com/lukemonahan/splunk_modinput_prometheus/issues/28). `curl` requests and desktop browsers (eg. Chrome) are served the whole body.

But on the other hand, this seems to be the only prometheus http service that client can't read from. It can read from node_exporter for example.

Any ideas?

By the way I'm not a Java programmer at all so anything 'obvious' like this or that -D flag etc may help a lot.

**Comment by akira-kurogane** [comment]
Yep, can try that (removing Accept-Encoding).

Here's the test's captured result.
```
akira@kfk-akira-1:~$ for stream in `tshark -r "$f" -2 -R "tcp and (http.request or http.response)" -T fields -e tcp.stream | sort -n | uniq`; do     echo "==========BEGIN REQUEST==========";     tshark -q -r "$f" -z follow,tcp,ascii,$stream;     echo "==========END REQUEST=========="; done;
==========BEGIN REQUEST==========

===================================================================
Follow: tcp,ascii
Filter: tcp.stream eq 0
Node 0: 10.146.0.6:36154
Node 1: 10.146.0.5:8297
150
GET /metrics HTTP/1.1
Host: kfk-akira-1.asia-northeast1-b.c.hsbc-stretched-cluster.internal:8297
User-Agent: Go-http-client/1.1
Junkheader: foo


	136
HTTP/1.1 200 OK
Date: Wed, 29 Jun 2022 16:04:20 GMT
Content-type: text/plain; version=0.0.4; charset=utf-8
Content-length: 143505


===================================================================
==========END REQUEST==========
```

So it seems the body was known.

I'm going to try a longer timeout in the client. It was 500ms.

**Comment by akira-kurogane** [comment]
... and it was the timing. The response came at ~560ms.

So, in the end in my case the browser and the curl command were not doing something crucially different with headers, or maybe something in the TCP transport options, but simply waiting longer.

And the alternative server I tested with (node_exporter) was presumably just faster.

**Comment by akira-kurogane** [comment]
Thanks, let's make it wrap.

**Comment by akira-kurogane** [comment]
Follow-up question; not a bug report just a question.
Is 500+ ms normal? Anything I can/should do to improve it?

The process I've added the agent to is a Kafka broker, currently idle. The config I'm using for jmx_exporter is below.
```yaml
lowercaseOutputName: true
rules:
- pattern : kafka.cluster<type=(.+), name=(.+), topic=(.+), partition=(.+)><>Value
  name: kafka_cluster_$1_$2
  labels:
    topic: "$3"
    partition: "$4"
- pattern : kafka.log<type=Log, name=(.+), topic=(.+), partition=(.+)><>Value
  name: kafka_log_$1
  labels:
    topic: "$2"
    partition: "$3"
- pattern : kafka.controller<type=(.+), name=(.+)><>(Count|Value)
  name: kafka_controller_$1_$2
- pattern : kafka.network<type=(.+), name=(.+)><>Value
  name: kafka_network_$1_$2
- pattern : kafka.network<type=(.+), name=(.+)PerSec, request=(.+)><>Count
  name: kafka_network_$1_$2_total
  labels:
    request: "$3"
- pattern : kafka.network<type=(.+), name=(\w+), networkProcessor=(.+)><>Count
  name: kafka_network_$1_$2
  labels:
    request: "$3"
  type: COUNTER
- pattern : kafka.network<type=(.+), name=(\w+), request=(\w+)><>Count
  name: kafka_network_$1_$2
  labels:
    request: "$3"
- pattern : kafka.network<type=(.+), name=(\w+)><>Count
  name: kafka_network_$1_$2
- pattern : kafka.server<type=(.+), name=(.+)PerSec\w*, topic=(.+)><>Count
  name: kafka_server_$1_$2_total
  labels:
    topic: "$3"
- pattern : kafka.server<type=(.+), name=(.+)PerSec\w*><>Count
  name: kafka_server_$1_$2_total
  type: COUNTER

- pattern : kafka.server<type=(.+), name=(.+), clientId=(.+), topic=(.+), partition=(.*)><>(Count|Value)
  name: kafka_server_$1_$2
  labels:
    clientId: "$3"
    topic: "$4"
    partition: "$5"
- pattern : kafka.server<type=(.+), name=(.+), topic=(.+), partition=(.*)><>(Count|Value)
  name: kafka_server_$1_$2
  labels:
    topic: "$3"
    partition: "$4"
- pattern : kafka.server<type=(.+), name=(.+), topic=(.+)><>(Count|Value)
  name: kafka_server_$1_$2
  labels:
    topic: "$3"
  type: COUNTER

- pattern : kafka.server<type=(.+), name=(.+), clientId=(.+), brokerHost=(.+), brokerPort=(.+)><>(Count|Value)
  name: kafka_server_$1_$2
  labels:
    clientId: "$3"
    broker: "$4:$5"
- pattern : kafka.server<type=(.+), name=(.+), clientId=(.+)><>(Count|Value)
  name: kafka_server_$1_$2
  labels:
    clientId: "$3"
- pattern : kafka.server<type=(.+), name=(.+)><>(Count|Value)
  name: kafka_server_$1_$2

- pattern : kafka.(\w+)<type=(.+), name=(.+)PerSec\w*><>Count
  name: kafka_$1_$2_$3_total
- pattern : kafka.(\w+)<type=(.+), name=(.+)PerSec\w*, topic=(.+)><>Count
  name: kafka_$1_$2_$3_total
  labels:
    topic: "$4"
  type: COUNTER
- pattern : kafka.(\w+)<type=(.+), name=(.+)PerSec\w*, topic=(.+), partition=(.+)><>Count
  name: kafka_$1_$2_$3_total
  labels:
    topic: "$4"
    partition: "$5"
  type: COUNTER
- pattern : kafka.(\w+)<type=(.+), name=(.+)><>(Count|Value)
  name: kafka_$1_$2_$3_$4
  type: COUNTER
- pattern : kafka.(\w+)<type=(.+), name=(.+), (\w+)=(.+)><>(Count|Value)
  name: kafka_$1_$2_$3_$6
  labels:
    "$4": "$5"
```

**Comment by fstab** [comment]
`jmx_exporter` queries JMX beans, and exposes the result as Prometheus metrics.

If querying the JMX beans takes long that's outside of `jmx_controller`'s control. JMX beans can do complex operations sometimes.

If the time is really spent in `jmx_exporter` code, i.e. formatting the result and exposing it in Prometheus format then we should try to figure out why and optimize it.

However, I don't think it's `jmx_exporter` code taking most of the time here.

**Comment by akira-kurogane** [comment]
Hmm, interesting. I don't know these programs at all really but I accept the simplest, and thus current best, hypothesis is the JMX MBeanInfo iteration, and how blocking that might be.

Before I used this I was fetching all the JMX MBeanInfo objects as  log lines of text and saw how much more space it needed in that log format. Much nicer for the destination db to use prometheus [metric,value] format.

Can I make iteration of MbeanInfo objects in JMX more efficient than my current config.yaml does? Eg. if I use _blacklistObjectNames_ or _whitelistObjectNames_ could that lead to the MBeans service doing less iteration? Or does it already only 'whitelist' nodes already matching the rules patterns.

**Comment by akira-kurogane** [comment]
Just wanted to share some visualized observations.

Using VisualVM I took at look at the CPU when jmx_exporter is enabled in an otherwise idle Kafka broker JVM and the prometheus output is request by a http client. At first I did it manually several times, then I set a collection process up to get it once per 10s.
![jmx_exporter_cpu_screenshot 20220701T00:42Z](https://user-images.githubusercontent.com/2595590/176801361-21d33053-7a82-4654-9f7a-50580283f7aa.jpg)

Given that this sampling is done once per second and we know the collection process in this JVM takes ~500ms a 50% peak = 100% CPU saturation, a 20% peak means 40% saturation, etc.
